### PR TITLE
Add some new feature

### PR DESCRIPTION
**Pull request by chteufleur**
Add some features about vCard, Ad-Hoc command and Software version.
Add also some new attribute into presence stanza and disco info.

### CODE REVIEW

File: src/xmpp/uuid.go [hunk 1]
```diff
@@ -15,3 +15,16 @@ func UUID4() string {
 	uuid[8] = (uuid[8] &^ 0x40) | 0x80
 	return fmt.Sprintf("%x-%x-%x-%x-%x", uuid[:4], uuid[4:6], uuid[6:8], uuid[8:10], uuid[10:])
 }
+
+func SessionId() string {
```
Comment (emgee) [inline src/xmpp/uuid.go]: I'd call this SessionID.

File: src/xmpp/uuid.go [hunk 2]
```diff
@@ -15,3 +15,16 @@ func UUID4() string {
...
+	var strSize = 15
+	var dictionary string
+	dictionary = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
```
Comment (emgee) [inline src/xmpp/uuid.go]: Can strSize and dictionary be consts?

File: src/xmpp/vcard.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+)
+
+const (
+	NsVCardTemp = "vcard-temp"
+)
+
+// XEP-0054 vCard
+type VCard struct {
+	XMLName xml.Name `xml:"vcard-temp vCard"`
+	// TODO Must complete truct
```
Comment (emgee) [inline src/xmpp/vcard.go]: This doesn't look very useful at the moment. Maybe drop it from the branch for now?

File: src/xmpp/ad-hoc.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	TypeAdHocFieldListSingle = "list-single"
+	TypeAdHocFieldTextSingle = "text-single"
+	TypeAdHocFieldJidSingle  = "jid-single"
+)
+
+type AdHocCommand struct {
+	XMLName   xml.Name   `xml:"http://jabber.org/protocol/commands command"`
+	Node      string     `xml:"node,attr"`
+	Action    string     `xml:"action,attr"`
+	SessionId string     `xml:"sessionid,attr"`
```
Comment (emgee) [inline src/xmpp/ad-hoc.go]: I'd call this SessionID.

File: src/xmpp/stanza.go [hunk 5]
```diff
@@ -6,6 +6,13 @@ import (
 	"fmt"
 )
 
+const (
+	IqTypeGet    = "get"
+	IqTypeSet    = "set"
+	IqTypeResult = "result"
+	IqTypeError  = "error"
```
Comment (emgee) [inline src/xmpp/stanza.go]: I think all of these should start with IQ.

File: src/xmpp/disco.go [hunk 6]
```diff
@@ -6,8 +6,8 @@ import (
 )
 
 const (
-	nsDiscoInfo  = "http://jabber.org/protocol/disco#info"
-	nsDiscoItems = "http://jabber.org/protocol/disco#items"
+	NsDiscoInfo  = "http://jabber.org/protocol/disco#info"
+	NsDiscoItems = "http://jabber.org/protocol/disco#items"
```
Comment (emgee) [inline src/xmpp/disco.go]: I think these should start with NS.

File: src/xmpp/uuid.go [hunk 7]
```diff
@@ -15,3 +15,16 @@ func UUID4() string {
...
+
+	var bytes = make([]byte, strSize)
+	rand.Read(bytes)
```
Comment (emgee) [inline src/xmpp/uuid.go]: This ignores errors.

File: src/xmpp/softwareVersion.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package xmpp
+
+import (
+	"encoding/xml"
+)
+
+const (
+	NsJabberClient = "jabber:iq:version"
```
Comment (emgee) [inline src/xmpp/softwareVersion.go]: I think this should be NSJabberClient.

File: src/xmpp/vcard.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package xmpp
+
+import (
+	"encoding/xml"
+)
+
+const (
+	NsVCardTemp = "vcard-temp"
```
Comment (emgee) [inline src/xmpp/vcard.go]: I think this should be NSVCardTemp.

**Comment by emgee** [comment]
Thanks for this. I added some comments (most minor).

**Comment by chteufleur** [comment]
I fixed all typo and SessionID.
But I think VCard is still usefull. It allow to reply vcard request like this:

`2016/05/16 21:10:28 recv: <iq type='get' to='component.example.com' from='user@example.com/Gajim' id='385' xmlns:='jabber:component:accept'><vCard xmlns='vcard-temp'/></iq>`
`2016/05/16 21:10:28 send: <iq id="385" type="result" to="user@example.com/Gajim" from="component.example.com"><vCard xmlns="vcard-temp"></vCard></iq>`

And it is the beginning for the most complete implementation.
If it's the TODO comment you don't like, I still can remove it.

**Comment by emgee** [comment]
Thanks for the changes. Sorry for taking so long to review this again, I'll try to get to it this weekend.

File: src/xmpp/dns.go [hunk 10]
```diff
@@ -27,7 +27,18 @@ func HomeServerAddrs(jid JID) (addr []string, err error) {
...
-		addr = append(addr, fmt.Sprintf("%s:%d", a.Target, a.Port))
+		target := parseTargetDomainName(a.Target)
+		addr = append(addr, fmt.Sprintf("%s:%d", target, a.Port))
+	}
+	return
+}
+
+// Remove the last dot in the domain name if exist
+func parseTargetDomainName(domainName string) (ret string) {
+	if domainName[len(domainName)-1] == '.' {
```
Comment (emgee) [inline src/xmpp/dns.go]: Out of interest, why is this necessary? But anyway, what about just `domainName = strings.TrimRight(domainName, ".")` instead of the slice and recursive call?

File: src/xmpp/stanza.go [hunk 11]
```diff
@@ -56,13 +67,29 @@ func (iq *Iq) Response(type_ string) *Iq {
...
+	Type    string        `xml:"type,attr,omitempty"`
+	To      string        `xml:"to,attr,omitempty"`
+	From    string        `xml:"from,attr,omitempty"`
+	Subject string        `xml:"subject,omitempty"`
+	Body    []MessageBody `xml:"body,omitempty"`
+	Thread  string        `xml:"thread,omitempty"`
+	Error   *Error        `xml:"error"`
+	Lang    string        `xml:"xml:lang,attr,omitempty"`
+
+	Confir *Confirm `xml:"confirm"` // XEP-0070
```
Comment (emgee) [inline src/xmpp/stanza.go]: s/Confir/Confirm ?

File: src/xmpp/stream.go [hunk 12]
```diff
@@ -16,6 +17,9 @@ type StreamConfig struct {
 	// are either sent to the server or delivered to the application. It also
 	// causes incoming stanzas to be XML-parsed a second time.
 	LogStanzas bool
+
+	// The dommain connection for certificat validation.
```
Comment (emgee) [inline src/xmpp/stream.go]: Typo - s/certificat/certificate

File: src/xmpp/dns.go (hunk 10 above)
Comment (chteufleur) [inline src/xmpp/dns.go]: If I remember well, it cause me some troubles, but can't reproduce. I didn't know about `strings.TrimRight()` which is better than what I've done. Will change that.

**Comment by emgee** [comment]
Thanks for the fixes and for running gofmt across the code. It looks good now ... I'll try to take a last look through over the weekend and then merge it.

File: src/xmpp/httpAuth.go [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+const (
+	NSHTTPAuth = "http://jabber.org/protocol/http-auth"
+)
+
+// XEP-0070: Verifying HTTP Requests via XMPP
+type Confirm struct {
+	XMLName xml.Name `xml:"http://jabber.org/protocol/http-auth confirm"`
+	Id      string   `xml:"id,attr"`
```
Comment (emgee) [inline src/xmpp/httpAuth.go]: This should be `ID` but it matches what I used already so let's leave it as `Id` for now and land this. However, I may fix the case of struct fields across the package - a breaking change, but worth it I think.
